File: src/ai/ml/encoding.ts
import {
  ROWS,
  COLS,
  AI_VIEW_ROWS,
  AI_ROW_OFFSET,
} from '../../game/constants';
import type { Color, GameState, Pair } from '../../game/types';
import { ACTION_COUNT, actionIndexToMove } from '../../game/action';

export const BOARD_CHANNELS = 11;
export const QUEUE_DIM = 16;
export const COLOR_ORDER: readonly Color[] = ['R', 'B', 'Y', 'P'];
const COLOR_INDEX: Record<Color, number> = { R: 0, B: 1, Y: 2, P: 3 };

export interface EncodedState {
  board: Float32Array;
  queue: Float32Array;
  legalMask: Uint8Array;
}

interface CanonState {
  field: { cells: (Color | null)[][] };
  current: GameState['current'];
  nextQueue: Pair[];
}

export function canonicalizeColors(state: GameState): {
  canonical: CanonState;
  perm: Partial<Record<Color, number>>;
} {
  const perm: Partial<Record<Color, number>> = {};
  const see = (c: Color | null | undefined) => {
    if (c == null) return;
    if (perm[c] === undefined && Object.keys(perm).length < 4) {
      perm[c] = Object.keys(perm).length;
    }
  };

  const cells = state.field.cells as readonly (readonly (Color | null)[])[];
  for (let r = ROWS - 1; r >= 0; r--) {
    for (let c = 0; c < COLS; c++) see(cells[r]![c]);
  }
  if (state.current) {
    see(state.current.pair.axis);
    see(state.current.pair.child);
  }
  for (const p of state.nextQueue) {
    see(p.axis);
    see(p.child);
  }

  const remap = (c: Color | null | undefined): Color | null => {
    if (c == null) return null;
    const id = perm[c];
    return id === undefined ? c : COLOR_ORDER[id]!;
  };

  const canonField: (Color | null)[][] = cells.map((row) =>
    row.map((c) => remap(c)),
  );
  const canonCurrent = state.current
    ? {
        ...state.current,
        pair: {
          axis: remap(state.current.pair.axis)!,
          child: remap(state.current.pair.child)!,
        },
      }
    : null;
  const canonQueue: Pair[] = state.nextQueue.map((p) => ({
    axis: remap(p.axis)!,
    child: remap(p.child)!,
  }));

  return {
    canonical: {
      field: { cells: canonField },
      current: canonCurrent,
      nextQueue: canonQueue,
    },
    perm,
  };
}

export function encodeState(state: GameState): EncodedState {
  const { canonical } = canonicalizeColors(state);
  // The trained policy/value net was built against AI_VIEW_ROWS (= 13). The
  // game now stores ROWS (= 14), with AI_ROW_OFFSET extra rows above the AI's
  // view. We drop those top rows when encoding so the tensor shape stays
  // [AI_VIEW_ROWS, COLS, BOARD_CHANNELS] and the existing model still loads.
  const board = new Float32Array(AI_VIEW_ROWS * COLS * BOARD_CHANNELS);
  const cellIdx = (r: number, c: number, ch: number) =>
    r * COLS * BOARD_CHANNELS + c * BOARD_CHANNELS + ch;
  // Translate an AI-view row index (0..AI_VIEW_ROWS-1) into the game-field
  // row index. The AI's row 0 corresponds to game row AI_ROW_OFFSET (= the
  // old "13段目" / ceiling).
  const fieldRow = (r: number) => r + AI_ROW_OFFSET;

  // ch 0..3 + ch 4 (empty)
  for (let r = 0; r < AI_VIEW_ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      const v = canonical.field.cells[fieldRow(r)]![c];
      if (v == null) {
        board[cellIdx(r, c, 4)] = 1;
      } else {
        board[cellIdx(r, c, COLOR_INDEX[v])] = 1;
      }
    }
  }

  // ch 5/6: axis/child broadcast
  if (canonical.current) {
    const ax = COLOR_INDEX[canonical.current.pair.axis] / 3;
    const ch = COLOR_INDEX[canonical.current.pair.child] / 3;
    for (let r = 0; r < AI_VIEW_ROWS; r++) {
      for (let c = 0; c < COLS; c++) {
        board[cellIdx(r, c, 5)] = ax;
        board[cellIdx(r, c, 6)] = ch;
      }
    }
  }

  // ch 7: heightmap (relative to AI_VIEW_ROWS so values stay in [0,1] like
  // training). Puyos sitting in the dropped top rows don't count toward
  // height in the AI's view, but they're rare and their effect on heuristics
  // is small.
  const heights = new Array<number>(COLS).fill(0);
  for (let c = 0; c < COLS; c++) {
    for (let r = 0; r < AI_VIEW_ROWS; r++) {
      if (canonical.field.cells[fieldRow(r)]![c] != null) {
        heights[c] = AI_VIEW_ROWS - r;
        break;
      }
    }
    for (let r = 0; r < AI_VIEW_ROWS; r++)
      board[cellIdx(r, c, 7)] = heights[c]! / AI_VIEW_ROWS;
  }

  // ch 8: 4-connected mask
  const mask = fourConnectedMask(canonical.field.cells);
  for (let r = 0; r < AI_VIEW_ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (mask[fieldRow(r)]![c]) board[cellIdx(r, c, 8)] = 1;
    }
  }

  // ch 9 / 10: ceiling and danger occupancy per column. In the AI's 13-row
  // view these refer to (AI) rows 0 and 1 — the topmost two visible/ceiling
  // rows. Map them back to the game field's corresponding rows.
  for (let c = 0; c < COLS; c++) {
    if (canonical.field.cells[fieldRow(0)]![c] != null) {
      for (let r = 0; r < AI_VIEW_ROWS; r++) board[cellIdx(r, c, 9)] = 1;
    }
    if (canonical.field.cells[fieldRow(1)]![c] != null) {
      for (let r = 0; r < AI_VIEW_ROWS; r++) board[cellIdx(r, c, 10)] = 1;
    }
  }

  // queue
  const queue = new Float32Array(QUEUE_DIM);
  if (canonical.nextQueue.length >= 1) {
    const n1 = canonical.nextQueue[0]!;
    queue[COLOR_INDEX[n1.axis]] = 1;
    queue[4 + COLOR_INDEX[n1.child]] = 1;
  }
  if (canonical.nextQueue.length >= 2) {
    const n2 = canonical.nextQueue[1]!;
    queue[8 + COLOR_INDEX[n2.axis]] = 1;
    queue[12 + COLOR_INDEX[n2.child]] = 1;
  }

  // legal mask
  const legalMask = new Uint8Array(ACTION_COUNT);
  if (canonical.current) {
    for (let i = 0; i < ACTION_COUNT; i++) {
      const m = actionIndexToMove(i);
      const dc = m.rotation === 1 ? 1 : m.rotation === 3 ? -1 : 0;
      if (
        m.axisCol >= 0 &&
        m.axisCol < COLS &&
        m.axisCol + dc >= 0 &&
        m.axisCol + dc < COLS
      ) {
        legalMask[i] = 1;
      }
    }
  }

  return { board, queue, legalMask };
}

function fourConnectedMask(cells: readonly (readonly (Color | null)[])[]): boolean[][] {
  const seen: boolean[][] = Array.from({ length: ROWS }, () =>
    Array<boolean>(COLS).fill(false),
  );
  const out: boolean[][] = Array.from({ length: ROWS }, () =>
    Array<boolean>(COLS).fill(false),
  );
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (seen[r]![c] || cells[r]![c] == null) continue;
      const color = cells[r]![c]!;
      const stack: [number, number][] = [[r, c]];
      const group: [number, number][] = [];
      while (stack.length) {
        const [y, x] = stack.pop()!;
        if (y < 0 || y >= ROWS || x < 0 || x >= COLS) continue;
        if (seen[y]![x] || cells[y]![x] !== color) continue;
        seen[y]![x] = true;
        group.push([y, x]);
        stack.push([y - 1, x], [y + 1, x], [y, x - 1], [y, x + 1]);
      }
      if (group.length >= 4) for (const [y, x] of group) out[y]![x] = true;
    }
  }
  return out;
}

File: python/puyo_train/encoding.py
from __future__ import annotations

import numpy as np

ROWS = 13
COLS = 6
BOARD_H = ROWS
BOARD_W = COLS
BOARD_CHANNELS = 11
QUEUE_DIM = 16
COLOR_ORDER = ("R", "B", "Y", "P")
_COLOR_INDEX = {c: i for i, c in enumerate(COLOR_ORDER)}


def encode_state(state: dict) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """state = {"field": [[color or None]], "current": {...}, "next_queue": [{...}]}

    Apply color canonicalization first, then build an 11ch board tensor:
      ch 0..3  one-hot per canonical color
      ch 4     empty cell
      ch 5     axis color (broadcast across all cells, scaled to [0,1])
      ch 6     child color (broadcast, scaled)
      ch 7     column heightmap (height / ROWS, broadcast across that column)
      ch 8     same-color 4-connected mask (1 where the cell already belongs
               to a 4-connected same-color group, else 0)
      ch 9     ceiling-row-occupied flag for the column (broadcast)
      ch 10    danger-row (row 1) occupied flag for the column (broadcast)
    """
    canon_state, _perm = canonicalize_colors(state)

    board = np.zeros((ROWS, COLS, BOARD_CHANNELS), dtype=np.float32)
    field = canon_state["field"]
    for r in range(ROWS):
        for c in range(COLS):
            cell = field[r][c]
            if cell is None:
                board[r, c, 4] = 1.0
            else:
                board[r, c, _COLOR_INDEX[cell]] = 1.0

    current = canon_state.get("current")
    if current is not None:
        ax = _COLOR_INDEX[current["axis"]] / 3.0
        ch = _COLOR_INDEX[current["child"]] / 3.0
        board[:, :, 5] = ax
        board[:, :, 6] = ch

    # ch 7: heightmap per column
    heights = _column_heights(field)
    for c in range(COLS):
        board[:, c, 7] = heights[c] / float(ROWS)

    # ch 8: 4-connected mask
    mask = _four_connected_mask(field)
    for r in range(ROWS):
        for c in range(COLS):
            if mask[r][c]:
                board[r, c, 8] = 1.0

    # ch 9: ceiling occupancy (row 0)
    for c in range(COLS):
        if field[0][c] is not None:
            board[:, c, 9] = 1.0
    # ch 10: danger row occupancy (row 1)
    for c in range(COLS):
        if field[1][c] is not None:
            board[:, c, 10] = 1.0

    queue = np.zeros((QUEUE_DIM,), dtype=np.float32)
    nq = canon_state.get("next_queue", [])
    if len(nq) >= 1:
        n1 = nq[0]
        queue[_COLOR_INDEX[n1["axis"]]] = 1.0
        queue[4 + _COLOR_INDEX[n1["child"]]] = 1.0
    if len(nq) >= 2:
        n2 = nq[1]
        queue[8 + _COLOR_INDEX[n2["axis"]]] = 1.0
        queue[12 + _COLOR_INDEX[n2["child"]]] = 1.0

    legal = _legal_mask(field, current)
    return board, queue, legal


def _column_heights(field) -> list[int]:
    heights = [0] * COLS
    for c in range(COLS):
        for r in range(ROWS):
            if field[r][c] is not None:
                heights[c] = ROWS - r
                break
    return heights


def _four_connected_mask(field) -> list[list[bool]]:
    """Mark cells that already belong to a same-color group of size >= 4."""
    seen = [[False] * COLS for _ in range(ROWS)]
    out = [[False] * COLS for _ in range(ROWS)]
    for r in range(ROWS):
        for c in range(COLS):
            if seen[r][c] or field[r][c] is None:
                continue
            color = field[r][c]
            stack = [(r, c)]
            group = []
            while stack:
                y, x = stack.pop()
                if (
                    y < 0 or y >= ROWS or x < 0 or x >= COLS
                    or seen[y][x] or field[y][x] != color
                ):
                    continue
                seen[y][x] = True
                group.append((y, x))
                stack.extend([(y - 1, x), (y + 1, x), (y, x - 1), (y, x + 1)])
            if len(group) >= 4:
                for y, x in group:
                    out[y][x] = True
    return out


def canonicalize_colors(state: dict) -> tuple[dict, dict[str, int]]:
    """Renames colors so that they appear in canonical order (R, B, Y, P) by
    first-appearance scan: field bottom→top + left→right, then current.axis,
    current.child, next1.axis, next1.child, next2.axis, next2.child.

    Returns (canonical_state, perm) where perm maps original color → canonical id.
    """
    perm: dict[str, int] = {}

    def _see(c):
        if c is None or c not in ("R", "B", "Y", "P"):
            return
        if c not in perm and len(perm) < 4:
            perm[c] = len(perm)

    field = state["field"]
    for r in range(12, -1, -1):
        for c in range(6):
            _see(field[r][c])

    cur = state.get("current")
    if cur is not None:
        _see(cur.get("axis"))
        _see(cur.get("child"))
    for pair in state.get("next_queue", []):
        _see(pair.get("axis"))
        _see(pair.get("child"))

    def _remap(c):
        if c is None or c not in perm:
            return c
        return COLOR_ORDER[perm[c]]

    canon_field = [[_remap(c) for c in row] for row in field]
    canon_state = {
        "field": canon_field,
        "current": (
            None
            if cur is None
            else {**cur, "axis": _remap(cur.get("axis")), "child": _remap(cur.get("child"))}
        ),
        "next_queue": [
            {**p, "axis": _remap(p.get("axis")), "child": _remap(p.get("child"))}
            for p in state.get("next_queue", [])
        ],
    }
    return canon_state, perm


def _legal_mask(field, current) -> np.ndarray:
    from .action import ACTION_COUNT, action_index_to_move

    mask = np.zeros((ACTION_COUNT,), dtype=np.uint8)
    if current is None:
        return mask
    for i in range(ACTION_COUNT):
        col, rot = action_index_to_move(i)
        dc = 1 if rot == 1 else -1 if rot == 3 else 0
        if 0 <= col < COLS and 0 <= col + dc < COLS:
            mask[i] = 1
    return mask
